packages/gatsby-recipes/src/renderer/render.js

```diff
@@ -202,43 +217,86 @@ const handleResource = (resourceName, context, props) => {
 
   inFlightCache.set(cacheKey, promise)
 
-  queue.add(() => promise)
+  queue.push(promise)
 
   throw promise
 }
 
-const render = async (recipe, cb, inputs = {}, isApply) => {
+const render = (recipe, cb, inputs = {}, isApply, isStream, name) => {
+  const emitter = mitt()
   const plan = {}
 
+  const queue = new Queue(
+    async (job, cb) => {
+      const result = await job
+      cb(null, result)
+    },
+    { concurrent: 5 }
+  )
+
+  queue.push(new Promise(resolve => resolve()))
+
+  const resultCache = new Map()
+  const inFlightCache = new Map()
+
+  let result
+
   const recipeWithWrapper = (
-    <Wrapper inputs={inputs} isApply={isApply}>
+    <Wrapper
+      inputs={inputs}
+      isApply={isApply}
+      resultCache={resultCache}
+      inFlightCache={inFlightCache}
+      queue={queue}
+    >
       {recipe}
     </Wrapper>
   )
 
-  const renderResources = async () => {
-    queue.pause()
+  // Keep calling render until there's remaining resources to render.
+  // This let's resources that depend on other resources to pause until one finishes.
+  const renderResources = isDrained => {
+    result = RecipesReconciler.render(recipeWithWrapper, plan, name)
+
+    // If there's still nothing on the queue that means we're done.
+    if (isDrained && queue.length === 0) {
+      // Rerender with the resources and resolve the data from the cache
+      result = RecipesReconciler.render(recipeWithWrapper, plan)
+      const resources = transformToPlan(result)
+      emitter.emit(`done`, resources)
+    }
+  }
 
-    RecipesReconciler.render(recipeWithWrapper, plan)
+  const throttledRenderResources = lodash.throttle(renderResources, 30, {
+    trailing: false,
+  })
 
-    if (!queue.size) {
-      return undefined
-    }
+  queue.on(`task_finish`, function (taskId, r, stats) {
+    throttledRenderResources()
 
-    queue.start()
-    await queue.onIdle()
-    return await renderResources()
-  }
+    const resources = transformToPlan(result)
+    emitter.emit(`update`, resources)
+  })
 
-  try {
-    // Begin the "render loop" until there are no more resources being queued.
-    await renderResources()
+  queue.on(`drain`, () => {
+    renderResources(true)
+  })
 
-    // Rerender with the resources and resolve the data from the cache
-    const result = RecipesReconciler.render(recipeWithWrapper, plan)
-    return transformToPlan(result)
-  } catch (e) {
-    throw e
+  renderResources()
+
+  if (isStream) {
+    return emitter
+  } else {
+    return new Promise((resolve, reject) => {
```

Comment:
default to the old promise resolve when things are done & let people opt into the stream interface